experimental/xformer/Transforms/OptimizeFullyConnected.cpp

```diff
@@ -26,13 +26,138 @@ struct OptimizeFullyConnected
 
 #include "Transforms/GeneratedPatterns.inc"
 
+struct LegalizeFullyConnected : public OpRewritePattern<FullyConnectedOp> {
+  using OpRewritePattern<FullyConnectedOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(FullyConnectedOp fcOp,
+                                PatternRewriter &rewriter) const override {
+
+    // TODO: Hack to prevent re-matching FullyConnectedOp for now
+    if (fcOp.keep_num_dims())
+      return failure();
+
+    // Get filter values as int64
+    auto filterQConstOp =
+        dyn_cast<TFL::QConstOp>(fcOp.filter().getDefiningOp());
+    DenseElementsAttr filterAttr =
+        filterQConstOp.value().cast<DenseElementsAttr>();
+    auto filterVector64 =
+        llvm::SmallVector<int64_t, 0>{filterAttr.getValues<int8_t>().begin(),
+                                      filterAttr.getValues<int8_t>().end()};
+
+    int p = 0;
+    for (auto const &i : filterVector64) {
+      llvm::errs() << p++ << " " << (int)i << "\n";
+    }
+
+    // Get input zero point
+    RankedTensorType inputType =
+        fcOp.input().getType().dyn_cast<RankedTensorType>();
+    auto inputQType = inputType.getElementType()
+                          .dyn_cast<mlir::quant::UniformQuantizedType>();
+    int64_t inputZeroPoint = inputQType.getZeroPoint();
+
+    // Multiply filter by input zero point and reduce on axis 1
+    // We will be left with "axis 0" number of elements in zeroPointBiasVector64
+    // which is the same as the number of outputs
+    llvm::SmallVector<int64_t, 0> zeroPointBiasVector64;
+    RankedTensorType filterType =
+        fcOp.filter().getType().dyn_cast<RankedTensorType>();
+    int axis1Size = filterType.getShape()[1];
+    int64_t sum = 0;
+    for (int i = 0; i < filterVector64.size(); ++i) {
+      sum = sum + filterVector64[i] * inputZeroPoint;
+      if (i % axis1Size == 0) {
+        zeroPointBiasVector64.push_back(sum);
+        sum = 0;
+      }
+    }
+
+    p = 0;
+    for (auto const &i : zeroPointBiasVector64) {
+      llvm::errs() << p++ << " " << (int)i << "\n";
+    }
+
+    // Get bias values as int64
+    auto biasQConstOp = dyn_cast<TFL::QConstOp>(fcOp.bias().getDefiningOp());
+    DenseElementsAttr biases = biasQConstOp.value().cast<DenseElementsAttr>();
+    auto biasVector64 = llvm::SmallVector<int64_t, 0>{
+        biases.getValues<int32_t>().begin(), biases.getValues<int32_t>().end()};
+
+    // Subtract to get unified bias
+    for (int i = 0; i < biasVector64.size(); ++i) {
+      biasVector64[i] = biasVector64[i] - zeroPointBiasVector64[i];
```

Comment:
```suggestion
      biasVector64[i] -= zeroPointBiasVector64[i];
```
Also, if you decide to use boost, you could use a `zip_iterator` (or something similar) to iterate through `biases` and `zeroPointBiasVector64` at the same time.